geom/twkb_parser.go

```diff
@@ -0,0 +1,528 @@
+}
+
+func (p *twkbParser) parseMultiPolygon() (MultiPolygon, error) {
+	if p.isEmpty {
+		return NewMultiPolygon(nil, p.opts...)
+	}
+	return p.nextMultiPolygon()
+}
+
+func (p *twkbParser) nextMultiPolygon() (MultiPolygon, error) {
+	numPolygons, err := p.parseUnsignedVarint()
+	if err != nil {
+		return MultiPolygon{}, fmt.Errorf("num polygons varint malformed: %w", err)
+	}
+	if err := p.parseIDList(int(numPolygons)); err != nil {
+		return MultiPolygon{}, err
+	}
+	var polys []Polygon
+	for i := 0; i < int(numPolygons); i++ {
+		poly, err := p.nextPolygon()
+		if err != nil {
+			return MultiPolygon{}, err
+		}
+		polys = append(polys, poly)
+	}
+	return NewMultiPolygon(polys, p.opts...)
+}
+
+func (p *twkbParser) parseGeometryCollection() (GeometryCollection, error) {
+	if p.isEmpty {
+		return NewGeometryCollection(nil, p.opts...), nil
+	}
+	return p.nextGeometryCollection()
+}
+
+func (p *twkbParser) nextGeometryCollection() (GeometryCollection, error) {
+	numGeoms, err := p.parseUnsignedVarint()
+	if err != nil {
+		return GeometryCollection{}, fmt.Errorf("num polygons varint malformed: %w", err)
+	}
+	if err := p.parseIDList(int(numGeoms)); err != nil {
+		return GeometryCollection{}, err
+	}
+	var geoms []Geometry
+	for i := 0; i < int(numGeoms); i++ {
+		subParser := newtwkbParser(p.twkb[p.pos:], p.opts...)
+		g, nbytes, err := subParser.parseGeometry()
+		if err != nil {
+			p.pos += nbytes // Add sub-parser's last known position, for error reporting.
+			return GeometryCollection{}, err
+		}
+		p.pos += nbytes // Sub-parser's geometry has been read, so ensure it is skipped.
+		geoms = append(geoms, g)
+	}
+	return NewGeometryCollection(geoms, p.opts...), nil
+}
+
+// Read a number of points then convert that many points from int to float coords.
+// Utilise and update the running memory of the previous reference point.
+// Return the slice of coords, the number of points, and any error.
+func (p *twkbParser) parsePointCountAndArray() ([]float64, int, error) {
+	numPoints, err := p.parseUnsignedVarint()
+	if err != nil {
+		return nil, 0, fmt.Errorf("num points varint malformed: %w", err)
+	}
+
+	coords, err := p.parsePointArray(int(numPoints))
+	return coords, int(numPoints), err
+}
+
+// Convert a given number of points from integer to floating point coordinates.
+// Utilise and update the running memory of the previous reference point.
+// The returned array will contain numPoints * the number of dimensions values.
+func (p *twkbParser) parsePointArray(numPoints int) ([]float64, error) {
+	var coords = make([]float64, numPoints*int(p.dimensions))
+	c := 0
+	for i := 0; i < numPoints; i++ {
+		for d := 0; d < p.dimensions; d++ {
+			val, err := p.parseSignedVarint()
+			if err != nil {
+				return nil, fmt.Errorf("point %d of %d coord %d varint malformed: %w", i, numPoints, d, err)
+			}
+
+			p.refpoint[d] += val // Reverse coord differencing to find the true value.
+			coords[c] = float64(p.refpoint[d]) * p.scalings[d]
```

Comment:
The conversion from the `p.refpoint[d]` value to `float64` is technically lossy above 2^52, since floats start "skipping" integer values above that point. I don't really see an easy way around it though... (e.g. we probably _don't_ want to mess around with arbitrary precision arithmetic). I don't think it will be a problem in practice, but thought I'd bring it up anyway in case you see an easy solution.

Reply:
Regarding converting coordinates to/from `float64`:

There are four bits in the TWKB header to indicate a signed integer for the power of 10 scaling factor for X and Y coordinates. The case we're probably most interested in is where we are keeping 7 decimal places in a latitude or longitude, which is the maximum number of decimal places TWKB lets us retain. Now 10^7 is approximately 2^24, and if we're keeping whole numbers up to +/-180 in front of that, we can assume another 8 bits on top, so it probably amounts to 32 bits of significant information for most real world cases.

My _feeling_ is the `int64` data we're dealing with won't have float conversion roundoffs, and probably won't lose much precision even after scaling. Besides, I don't see a good way around a hypothetical float roundoff, it's simply inherent to the efficiency of using floats.